File: leetcode/101-200/124-binary-tree-maximum-path-sum/bench/max_path_sum.c
/* LeetCode #124 benchmark mirror (C) — raw-pointer binary tree, the metal floor.
 * Same algorithm and tree construction as the Kara/Rust/Go/Python mirrors. */
#include <stdio.h>
#include <stdlib.h>

#define TREE_COUNT 2048
#define NODE_COUNT 511
#define REPS 60

typedef struct TreeNode {
    long val;
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;

static long node_value(long i, long seed) {
    return ((i * 37 + seed * 13) % 41) - 20;
}

static TreeNode *build_balanced(long lo, long hi, long seed) {
    if (lo > hi) return NULL;
    long mid = (lo + hi) / 2;
    TreeNode *node = (TreeNode *)malloc(sizeof(TreeNode));
    node->val = node_value(mid, seed);
    node->left = build_balanced(lo, mid - 1, seed);
    node->right = build_balanced(mid + 1, hi, seed);
    return node;
}

static long max_gain(TreeNode *node, long *best) {
    if (node == NULL) return 0;
    long lg = max_gain(node->left, best);
    long rg = max_gain(node->right, best);
    long left_gain = lg > 0 ? lg : 0;
    long right_gain = rg > 0 ? rg : 0;
    long through = node->val + left_gain + right_gain;
    if (through > *best) *best = through;
    long branch = left_gain > right_gain ? left_gain : right_gain;
    return node->val + branch;
}

static long max_path_sum(TreeNode *root) {
    long best = -1000000000L;
    max_gain(root, &best);
    return best;
}

int main(void) {
    TreeNode **forest = (TreeNode **)malloc(sizeof(TreeNode *) * TREE_COUNT);
    for (long t = 0; t < TREE_COUNT; t++)
        forest[t] = build_balanced(0, NODE_COUNT - 1, t + 1);

    long sink = 0;
    for (long k = 0; k < REPS; k++)
        for (long i = 0; i < TREE_COUNT; i++)
            sink += max_path_sum(forest[i]);

    printf("%ld\n", sink);
    return 0;
}

File: leetcode/101-200/124-binary-tree-maximum-path-sum/bench/max_path_sum.rs
// LeetCode #124 benchmark mirror (Rust). Kara's semantic peer: an RC tree
// (`Option<Rc<TreeNode>>`) mirroring the `shared struct TreeNode`, read-only
// post-order traversal. Same algorithm and tree construction as the other mirrors.
use std::rc::Rc;

const TREE_COUNT: i64 = 2048;
const NODE_COUNT: i64 = 511;
const REPS: i64 = 60;

struct TreeNode {
    val: i64,
    left: Option<Rc<TreeNode>>,
    right: Option<Rc<TreeNode>>,
}

fn node_value(i: i64, seed: i64) -> i64 {
    ((i * 37 + seed * 13) % 41) - 20
}

fn build_balanced(lo: i64, hi: i64, seed: i64) -> Option<Rc<TreeNode>> {
    if lo > hi {
        return None;
    }
    let mid = (lo + hi) / 2;
    Some(Rc::new(TreeNode {
        val: node_value(mid, seed),
        left: build_balanced(lo, mid - 1, seed),
        right: build_balanced(mid + 1, hi, seed),
    }))
}

fn max_gain(node: &Option<Rc<TreeNode>>, best: &mut i64) -> i64 {
    match node {
        None => 0,
        Some(n) => {
            let lg = max_gain(&n.left, best);
            let rg = max_gain(&n.right, best);
            let left_gain = if lg > 0 { lg } else { 0 };
            let right_gain = if rg > 0 { rg } else { 0 };
            let through = n.val + left_gain + right_gain;
            if through > *best {
                *best = through;
            }
            let branch = if left_gain > right_gain { left_gain } else { right_gain };
            n.val + branch
        }
    }
}

fn max_path_sum(root: &Option<Rc<TreeNode>>) -> i64 {
    let mut best: i64 = -1000000000;
    max_gain(root, &mut best);
    best
}

fn main() {
    let forest: Vec<Option<Rc<TreeNode>>> =
        (0..TREE_COUNT).map(|t| build_balanced(0, NODE_COUNT - 1, t + 1)).collect();

    let mut sink: i64 = 0;
    for _ in 0..REPS {
        for root in &forest {
            sink += max_path_sum(root);
        }
    }
    println!("{}", sink);
}
